### PR TITLE
[WIP][SYCL][FPGA] Runtime implementation for host_pipes

### PR DESCRIPTION
**Pull request by sherry-yuan**
To align with spec: https://github.com/intel/llvm/pull/5838

Host pipe is a FIFO construct that provide links between elements of design that are accessed through read/write API. Host pipe is a pipe that links device kernel with host program.

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -2663,6 +2672,16 @@ class __SYCL_EXPORT handler {
   /// The list of valid SYCL events that need to complete
   /// before barrier command can be executed
   std::vector<detail::EventImplPtr> MEventsWaitWithBarrier;
+  /// Host pipe name
+  std::string HostPipeName;
+  /// Host pipe host pointer
```
Comment (keryell) [inline sycl/include/CL/sycl/handler.hpp:2677]: Great to have comments!
But they could be more precise than just reading the variable name. I have no idea about what is the host pointer.

File: sycl/include/sycl/ext/intel/experimental/host_pipes.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  // Non-blocking pipes
+  static _dataT read(queue & q, bool &success_code,
+                     memory_order order = memory_order::seq_cst);
+  static void write(queue & q, const _dataT &data, bool &success_code,
+                    memory_order order = memory_order::seq_cst);
+
+private:
+  static constexpr int32_t m_Size = sizeof(_dataT);
+  static constexpr int32_t m_Alignment = alignof(_dataT);
+  static constexpr int32_t ID = _name::id;
```
Comment (keryell) [inline sycl/include/sycl/ext/intel/experimental/host_pipes.hpp:104]: What if the programmer use a
```c++
#define ID 42
```
?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -7577,6 +7577,62 @@ pi_result piextUSMGetMemAllocInfo(pi_context Context, const void *Ptr,
   return PI_SUCCESS;
 }
 
+/// Host Pips
```
Comment (keryell) [inline sycl/plugins/level_zero/pi_level_zero.cpp:7580]: Pipes?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -7577,6 +7577,62 @@ pi_result piextUSMGetMemAllocInfo(pi_context Context, const void *Ptr,
...
+/// API to read host pipe
+/// @param queue a valid host command-queue in which the read / write command
+/// will be queued. command_queue and program must be created with the same
+/// OpenCL context.
+/// @param program a program object with a successfully built executable.
+/// @param pipe_symbol the name of the program scope pipe global variable.
+/// @param blocking indicate if the read and write operations are blocking or
+/// non-blocking
+/// @param ptr a pointer to buffer in host memory where data is to be read into
+/// or written from.
```
Comment (keryell) [inline sycl/plugins/level_zero/pi_level_zero.cpp:7591]: From which perspective?

**Review by keryell — COMMENTED** [review]
Interesting

File: sycl/include/sycl/ext/intel/experimental/host_pipes.hpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+template <class _name, typename _dataT,
+          typename PropertyList = default_pipe_properties>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable("sycl-host-access",
+                                                         "readwrite")]]
+#endif
+    host_pipe { // TODO change name to pipe, and merge into the existing pipe
+                // implementation
+  static_assert(
```
Comment (sherry-yuan) [inline sycl/include/sycl/ext/intel/experimental/host_pipes.hpp:77]: make a type trait instead

File: sycl/source/detail/host_pipe.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  const device Dev = q.get_device();
+  bool IsReadPipeSupported =
+      Dev.has_extension("cl_intel_program_scope_host_pipe");
+  if (!IsReadPipeSupported) {
+    return &_dataT();
+  }
+  // TODO: Get pipe name from template, get host pointer by quering the host
+  // pipe registration / host pipe mapping
+  _dataT data;
+  const std::string pipe_name = std::to_string(ID);
```
Comment (sherry-yuan) [inline sycl/source/detail/host_pipe.cpp:29]: Change to _get_built_in_pipe_id

File: sycl/source/detail/scheduler/commands.cpp [hunk 7]
```diff
@@ -2118,6 +2118,39 @@ cl_int enqueueImpKernel(
...
+                                 const std::string &PipeName, bool blocking,
+                                 void *ptr, size_t size,
+                                 std::vector<RT::PiEvent> &RawEvents,
+                                 RT::PiEvent *OutEvent, bool read) {
+  detail::OSModuleHandle M =
+      detail::OSUtil::getOSModuleHandle("HostPipeReadWriteKernelName");
+  RT::PiProgram Program =
+      sycl::detail::ProgramManager::getInstance().getBuiltPIProgram(
+          M, Queue->getContextImplPtr(), Queue->getDeviceImplPtr(), KernelName);
+
```
Comment (sherry-yuan) [inline sycl/source/detail/scheduler/commands.cpp:2132]: Only need get the program base on the pipe type, require frontend to provide the info.

**Comment by sherry-yuan** [comment]
closing because superseded by #5894

**Pull request closed without merging**